### ISSUE TITLE
ClientResponseFilter not called on UnknownHostException

### ISSUE THREAD

**Issue by pavolloffay** [eclipse-ee4j/jaxrs-api#684]
At the moment there is no way how to intercept exceptions like `UnknownHostException` (maybe also `RedirectionException`) in `ClientResponseFilter`. Note that `ClientResponseFilter` is called for a such request. If the integration uses both filters and response filter is used to free up resources this leads to leaks.

**Comment by mkarg** [comment]
I assume you meant to say *Note that Client**Request**filter is called for such a request*.

Your approach is not specification compliant. According to JAX-RS 2.1 chapter 6.1 `...a ClientResponseFilter is executed upon receiving a server response...`. In the case of `UnknownHostException` no response is actually received, hence it is obvious and correct that no response filter is invoked.

You need to attach your resources to the request object (as a request property), so they become eligible for GC once the request is cancelled. If this is not enough (e. g. too lazy), we need to extend a future version of the JAX-RS specification in a way that allows you to get an active callback on request cancellation. Is that what you strive for with this issue? In the end it would be solved by the proposal we already discussed shortly in a different thread: A custom `Extension` available on the client's Class-Path could be auto-discovered invoked before and and after talking HTTP, so it would be the right place to set up and tear down resources. But that was not wanted by some people, as we plan to add CDI SE support for JAX-RS 3.0 anyways.

**Comment by pavolloffay** [comment]
Yes I meant `ClientRequestfilter`. I do not have a strong opinion how this should be done but I think jax-rs should provide an API to intercept all calls, even the failed ones.

> we need to extend a future version of the JAX-RS specification in a way that allows you to get an active callback on request cancellation

Yes that is what I mean. We need a callback which is called _always_ if the request filter was called.

I think we could go even further and have an interceptor per network attempt e.g. if the jax-rs supports redirects.

**Comment by najibk** [comment]
Is there any known workarounds or implementation to achieve this ? I would also like to intercept i/o exceptions in a centralized manner.

**Comment by jansupol** [comment]
Jersey supports a [PostInvocationInterceptor](https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest3x/client.html#d0e5354) that gathers the exceptions, including `UnknownHostException`. It is a proprietary solution, though.

Keep in mind that the MP REST Client will become a part of Jakarta REST and any changes in the client API should be compatible with it.

**Comment by jonatan-ivanov** [comment]
Hi,
Were there any updates on this? Can JAX-RS provide a similar mechanism to what Jersey has (see https://github.com/jakartaee/rest/issues/684#issuecomment-1048880739)?

**Comment by mkarg** [comment]
Unfortunately I need to tell you that so far the vendors backing JAX-RS did not agree upon a cross-vendor API for this use case.

@jamezp @jansupol @jim-krueger WDYT?

**Comment by jamezp** [comment]
I'm personally a +1 to waiting for the CDI integration to become defined so CDI interceptors can be used. That said, I don't see why you can't already use an interceptor as CDI, in containers at least, is supported.

If we're talking about a client side that runs outside of a container, I could see another API being useful. However, I think we need to decide what will happen with clients and CDI integration.
